### PR TITLE
Flood metrics computation

### PR DESCRIPTION
**Pull request by giltinde**
It's a big one

### CODE REVIEW

File: databricks/flood-metrics-computation.py
Comment (giltinde) [inline databricks/flood-metrics-computation.py:1]: Main file of interest

File: flood/spark/transforms.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    """
+    Round a value to the given number of decimal places.
+    """
+    return round(value, precision)
+
+def create_round_udf(precision=3):
+    """
+    Create a UDF for rounding to the specified precision.
+    """
+    return F.udf(lambda x: round_to_precision(x, precision), DoubleType())
```
Comment (A-Stangeland) [inline flood/spark/transforms.py:15]: I would remove this and replace the usage with `F.round`.

**Review by A-Stangeland — COMMENTED** [review]
Looks good! Made some small sugestions.

File: databricks/flood-metrics-computation.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+# although it is assumed to already have been
+# done in the threshold joining operation
+threshold_df = spark.read.parquet(threshold_file_path)\
+                         .withColumn("latitude", round_udf("latitude"))\
+                         .withColumn("longitude", round_udf("longitude"))
+
+# COMMAND ----------
+
+# Broadcast thresholds as it is joined to forecast dataframe
+# threshold_df = F.broadcast(threshold_df)
```
[the remark this answers is not in the document]
Comment (giltinde) [reply databricks/flood-metrics-computation.py:158]: Done

File: flood/spark/transforms.py (hunk 1 above)
Comment (giltinde) [reply flood/spark/transforms.py]: Done

File: databricks/flood-metrics-computation.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+# Read and broadcast the threshold dataframe
+threshold_file_path = os.path.join(DBUTILS_PREFIX, S3_GLOFAS_AUX_DATA_PATH, 
+                                   GLOFAS_PROCESSED_THRESH_FILENAME)
+# Round the lat/lon columns as a safety measure
+# although it is assumed to already have been
+# done in the threshold joining operation
+threshold_df = spark.read.parquet(threshold_file_path)\
+                         .withColumn("latitude", round_udf("latitude"))\
+                         .withColumn("longitude", round_udf("longitude"))
```
[the remark this answers is not in the document]
Comment (giltinde) [reply databricks/flood-metrics-computation.py:153]: Done

File: databricks/flood-metrics-computation.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+# MAGIC
+# MAGIC **Read the GloFAS forecast data**
+
+# COMMAND ----------
+
+# Load all the forecast data from a folder into a single dataframe
+all_forecasts_df = spark.read.schema(CustomSchemaWithoutTimestamp)\
+                        .parquet(processed_discharge_filepath)\
+                        .withColumn("latitude", round_udf("latitude"))\
+                        .withColumn("longitude", round_udf("longitude"))\
```
[the remark this answers is not in the document]
Comment (giltinde) [reply databricks/flood-metrics-computation.py:132]: Done

File: databricks/flood-metrics-computation.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+# MAGIC %md
+# MAGIC
+# MAGIC **Defining paths**
+
+# COMMAND ----------
+
+# Create folder on cloud storage
+target_folder = os.path.join(DBUTILS_PREFIX, S3_GLOFAS_PROCESSED_PATH, formatted_date)
+dbutils.fs.mkdirs(target_folder)
```
[the remark this answers is not in the document]
Comment (giltinde) [reply databricks/flood-metrics-computation.py:250]: Good idea!

File: databricks/flood-api-examples.py
Comment (giltinde) [inline databricks/flood-api-examples.py:1]: Have another look at this updated version of the suggested way to extract data based on a query with lat/lon coordinates. Neighboring cells can be extracted even if the primary cell (the queried one) doesn't have a flood forecast for it.

**Review by A-Stangeland — APPROVED** [review]

**Pull request merged**